src/main/java/no/unit/pnxservice/Pnxervices.java

```diff
@@ -0,0 +1,153 @@
+    private static final String PNX_B_TITLE_KEY = "btitle";
+    private static final String PNX_LAD11_KEY = "lad11";
+    private static final String PNX_DISPLAY_KEY = "display";
+    private static final String PNX_CREATION_DATE_KEY = "creationdate";
+    private static final String PNX_FACETS_KEY = "facets";
+    private static final String PNX_LIBRARY_KEY = "library";
+
+    //output json key names:
+    private static final String EXTRACTED_RECORD_ID_KEY = "record_id";
+    private static final String EXTRACTED_SOURCE_KEY = "source";
+    private static final String EXTRACTION_PUBLICATION_PLACE_KEY = "publication_place";
+    private static final String EXTRACTED_MMS_ID_KEY = "mms_id";
+    private static final String EXTRACTED_B_TITLE_KEY = "b_title";
+    private static final String EXTRACTED_CREATION_YEAR_KEY = "creation_year";
+    private static final String EXTRACTED_DISPLAY_TITLE_KEY = "display_title";
+    private static final String EXTRACTED_LIBRARIES_KEY = "libraries";
+
+    //json key names that both output json and pnx json share:
+    private static final String ISBN = "isbn";
+    private static final String VOLUME = "volume";
+    private static final String PAGES = "pages";
+    private static final String CREATOR = "creator";
+    private static final String TITLE = "title";
+    private static final String PUBLISHER = "publisher";
+
+
+
+
+    private final transient PnxServiceConnection connection;
+
+
+    /**
+     *contructor.
+     */
+    public Pnxervices() {
+        this.connection = new PnxServiceConnection();
+    }
+
+    public Pnxervices(PnxServiceConnection connection) {
+        this.connection = connection;
+    }
+
+
+    /**
+     * copied from old version.
+     * @param docID used for retrieving the PNX record
+     * @return docID without primoRecordPrefix
+     */
+    protected String removePrimoRecordPrefix(String docID) {
+        return docID.replaceFirst(PRIMO_RECORD_PREFIX, "");
+    }
+
+    public JsonObject getPnxData(String documentId) {
+        JsonObject fullPNX = getFullPNX(documentId);
+        return extractUsefulDataFromXservice(fullPNX);
+    }
+
+
+
+
+    protected JsonObject getFullPNX(String documentId) {
+        String docID = removePrimoRecordPrefix(documentId);
+        try (InputStreamReader streamReader = connection.connect(docID)) {
+            String json = new BufferedReader(streamReader)
+                    .lines()
+                    .collect(Collectors.joining(System.lineSeparator()));
+            return JsonParser.parseString(json).getAsJsonObject();
+        } catch (URISyntaxException e) {
+            log.error(WRONG_URL_FOR_PRIMO_API, documentId, e);
+        } catch (IOException e) {
+            log.error(ERROR_WHILE_GETTING_AT_PRIMO_API_FOR,  documentId, e);
+        }
+        return new JsonObject();
+    }
+
+    protected JsonObject extractUsefulDataFromXservice(JsonObject response) {
```

Comment:
Jeg personlig er ikke en fan av JsonObject som return type. Jeg foretrekker String, eller enda mer en ordentlig Object. Something along the lines PnxBean eller IllTransferMessage eller InterLibraryLoanDTO. Da vi sender den ut i GatewayResponse gjør vi uansett en .toString().

Reply:
fikset